File: apps/backend/app/auth/credential_store.py
import logging
from typing import Optional, Dict

logger = logging.getLogger("credential-store")

class CredentialStore:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(CredentialStore, cls).__new__(cls)
            cls._instance.store: Dict[str, Dict[str, str]] = {}
        return cls._instance

    async def set_credential(self, project_id: str, username: str, password: str) -> None:
        """Stores credentials for a project."""
        logger.info(f"Saving credentials in CredentialStore for project {project_id}")
        self.store[project_id] = {
            "username": username,
            "password": password
        }

    async def get_credential(self, project_id: str) -> Optional[Dict[str, str]]:
        """Retrieves stored credentials for a project, falling back to default test credentials."""
        logger.info(f"Retrieving credentials from CredentialStore for project {project_id}")
        if project_id in self.store:
            return self.store[project_id]

        return {
            "username": "testuser@example.com",
            "password": "SecureTestPassword123!"
        }

credential_store = CredentialStore()

File: packages/agents/src/auth/credential-store.ts
import type { CredentialRef } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";
import { getDb, projects, eq } from "@testpilot/database";

const logger = createLogger("credential-store");

export interface DecryptedCredential {
  username?: string;
  password?: string;
  secretToken?: string;
}

/**
 * CredentialStore Abstraction.
 * Decouples raw secret/password retrieval from database storage.
 * Easily pluggable into AWS Secrets Manager, HashiCorp Vault, or Azure Key Vault.
 */
export class CredentialStore {
  private static instance: CredentialStore;

  private constructor() {}

  public static getInstance(): CredentialStore {
    if (!CredentialStore.instance) {
      CredentialStore.instance = new CredentialStore();
    }
    return CredentialStore.instance;
  }

  /** Retrieve and decrypt credential secrets for a given project */
  public async getCredential(projectId: string): Promise<DecryptedCredential | null> {
    logger.info({ projectId }, "Retrieving project credentials from CredentialStore");

    try {
      const db = getDb();
      const [project] = await db.select().from(projects).where(eq(projects.id, projectId));

      if (!project || (!project.testEmail && !project.testPassword)) {
        return null;
      }

      return {
        username: project.testEmail || undefined,
        password: project.testPassword || undefined,
      };
    } catch (err) {
      logger.error({ projectId, err }, "Failed to fetch credentials from store");
      return null;
    }
  }
}

export const credentialStore = CredentialStore.getInstance();
